Button component added

igorls opened the pull request:
New class Material.Button according to the guidelines, adjustments are still needed. Current options are Raised/Flat, Enabled/Disabled, Background and Ink (ripple) colors.

Demo added to the MainView.ux

Also adding Roboto default font:
Default Material typeface should be Roboto per the guidelines, changed Material.Text class to comply with https://material.io/guidelines/style/typography.html

eksperts commented on Components/Button.ux:
```diff
@@ -0,0 +1,71 @@
+<Panel ux:Class="Material.Button" Alignment="Center" Enabled="true" Raised="true" Height="48" MinWidth="88" Padding="8,0" HitTestMode="LocalBoundsAndChildren" BackgroundColor="White" InkColor="#ccc">
```
I would use `HitTestMode="LocalBounds"` instead, that seems sufficient

igorls replied:
fixed

eksperts commented on Components/Button.ux:
```diff
@@ -0,0 +1,71 @@
+	<float4 ux:Property="BackgroundColor"/>
+	<float4 ux:Property="InkColor"/>
+
+	<Rectangle ux:Name="RaisedButton_body" ZOffset="1" Height="36" Padding="8,0" CornerRadius="2">
```
When assigning a `ux:Name` to a node, the value should start with a lowercase letter and be in camelCase. This here, for example, should be `ux:Name="raisedButtonBody"`.

eksperts replied:
And more of a question rather than a request for changes - is the ZOffset necessary? Can't you instead reorder the nodes so that the ones on top would be on top of the others in UX file?

igorls replied:
I reordered, but the ZOffset is necessary actually

eksperts replied:
Why is it necessary? I don't see any triggers involved in manipulating them, so I don't immediately see the reason.

eksperts replied:
@igorls I tested and the `ZOffset`s do not seem to be necessary at all. You only need to put the nodes in the right order, just like layers in photoshop - the ones on top in UX are rendered on top of others.

eksperts commented on Components/Button.ux:
```diff
@@ -0,0 +1,71 @@
+	<float4 ux:Property="InkColor"/>
+
+	<Rectangle ux:Name="RaisedButton_body" ZOffset="1" Height="36" Padding="8,0" CornerRadius="2">
+		<SolidColor Color="{ReadProperty this.BackgroundColor}" />
```
You don't need to prefix these properties with an explicit `this.*`, since they are referring to the whole `ux:Class` we are currently in and `this.*` is implied.

eksperts commented on Components/Button.ux:
```diff
@@ -0,0 +1,71 @@
+
+	<Rectangle ux:Name="RaisedButton_body" ZOffset="1" Height="36" Padding="8,0" CornerRadius="2">
+		<SolidColor Color="{ReadProperty this.BackgroundColor}" />
+		<WhileTrue Value="{ReadProperty this.Raised}">
```
You don't need to prefix these properties with an explicit `this.*`, since they are referring to the whole `ux:Class` we are currently in and `this.*` is implied.

eksperts commented on Components/Button.ux:
```diff
@@ -0,0 +1,71 @@
+	</Rectangle>
+
+	<Rectangle ux:Name="innerPanel" ZOffset="3" Height="{ReadProperty RaisedButton_body.Height}" Padding="{ReadProperty RaisedButton_body.Padding}" CornerRadius="{ReadProperty RaisedButton_body.CornerRadius}" ClipToBounds="true">
+		<Text Value="{ReadProperty this.Text}" Font="RobotoMedium" ux:Name="text" Alignment="Center" FontSize="14" />
```
You don't need to prefix these properties with an explicit `this.*`, since they are referring to the whole `ux:Class` we are currently in and `this.*` is implied.

eksperts replied:
`ux:Name` should always be the first property on a node, followed by `Value` in case of `<Text />`.

eksperts commented on Components/Button.ux:
```diff
@@ -0,0 +1,71 @@
+		<Text Value="{ReadProperty this.Text}" Font="RobotoMedium" ux:Name="text" Alignment="Center" FontSize="14" />
+	</Rectangle>
+
+	<WhileTrue Value="{ReadProperty this.Enabled}" Invert="true">
```
You don't need to prefix these properties with an explicit `this.*`, since they are referring to the whole `ux:Class` we are currently in and `this.*` is implied.

eksperts replied:
Instead of using `<WhileTrue Invert="true">` can't you use a plain `<WhileFalse>`?

eksperts commented on Components/Button.ux:
```diff
@@ -0,0 +1,71 @@
+
+	</WhileTrue>
+
+	<WhileTrue Value="{ReadProperty this.Enabled}">
```
You don't need to prefix these properties with an explicit `this.*`, since they are referring to the whole `ux:Class` we are currently in and `this.*` is implied.

eksperts commented on Components/Button.ux:
```diff
@@ -0,0 +1,71 @@
+				<Set scaleAnimator.Duration="0.3" AtProgress="0" />
+				<Set showInk.Value="true" AtProgress="0" />
+				<Set showInk.Value="false" AtProgress="1" />
+				<Set scaleAnimator.Duration="0.6" AtProgress="1" />
```
It would be nice if these triggers were grouped together by the `AtProgress` value (0/1), otherwise it's hard to read what happens when.

eksperts commented on MainView.ux:
```diff
@@ -3,14 +3,27 @@
 			</Material.AppBar>
+
+			<Rectangle Dock="Top" Margin="5">
+				<StackLayout/>
```
Why use a `Rectangle` with a `StackLayout` when you can just have a `StackPanel`?

igorls replied:
I just wanted to use a stroke to separate the button example area from other upcoming examples

eksperts replied:
That still should be done differently though. Go with a `StackPanel` and we'll figure it out from there.

igorls commented:
all requested changes were fixed, exception to the ZOffsets that were actually needed.

igorls committed:
Adding Roboto default font

Default Material typeface should be Roboto per the guidelines, changed Material.Text class to comply (https://material.io/guidelines/style/typography.html)

igorls committed:
Added Material.Button (work in progress)

New class Material.Button according to the guidelines, adjustments are still needed. Current options are Raised/Flat Enabled/Disabled, as long as Background and Ink (ripple) colors.

Demo added to the MainView.ux

igorls committed:
fix

igorls committed:
small adjustments according to requested changes

Duckers commented on Style/Fonts.ux:
```diff
@@ -0,0 +1,5 @@
+<Panel ux:Class="Material.Fonts">
```
Please don't add these fonts, it severely bloats the app. 

This theme will use the platform's default font, which is Roboto on Android and San Francisco on iOS. 

Bundling full Roboto will add 10MB to the app.

igorls replied:
Ok, no problem, I also had a doubt about this

eksperts commented on MainView.ux:
```diff
@@ -3,14 +3,27 @@
+				</StackPanel>
+				<StackPanel Orientation="Horizontal">
+					<Material.Button Text="Flat Button - Enabled" Raised="false" Enabled="true" BackgroundColor="#FFFFFF00" InkColor="Material.Blue300"/>
+					<Material.Button Text="Flat Button - Disabled" Raised="false" Enabled="false" BackgroundColor="#FFFFFF00"/>
```
Is the background on these fully transparent on purpose? Wouldn't you want to reuse the same color as on the raised buttons, `Material.Blue200`?

igorls replied:
I chose to put no background on those to show that the ink effect applies to the area in case of a flat button, what do you think?

igorls committed:
Reverting to default fonts as requested

igorls commented:
I think it's better for us to hold that PR until I replace my own implementation of the ink effect with the same one @Duckers has used in the other components, main reason being the glitch with the CornerRadius non-zero. Also to focus all effects on the same uno class, avoiding problems in the future
